Python/Module6_Testing/Pytest.md

```diff
@@ -0,0 +1,750 @@
+
+```python
+@pytest.mark.parametrize("<param-name>", [<val-1>, <val-2>, ...])
+def test_function(<param-name>):
+    ...
+```
+
+We will often have tests that require multiple parameters.
+The general form for creating the the parameterization decorator for $N$ parameters,
+each of which assume $J$ values, is:
+
+```python
+@pytest.mark.parametrize("<param-name1>, <param-name2>, [...], <param-nameN>", 
+                         [(<param1-val1>, <param2-val1>, [...], <paramN-val1>),
+                          (<param1-val2>, <param2-val2>, [...], <paramN-val2>),
+                          ...
+                          (<param1-valJ>, <param2-valJ>, [...], <paramN-valJ>),
+                         ])
+def test_function(<param-name1>, <param-name2>, [...], <param-nameN>):
+    ...
+```
+
+For example, let's take the following trivial test:
+
+```python
+def test_inequality_unparameterized():
+    assert 1 < 2 < 3
+    assert 4 < 5 < 6
+    assert 7 < 8 < 9
+    assert 10 < 11 < 12
+```
+
+and rewrite it in parameterized form. 
+The decorator will have three distinct parameter, and each parameters, let's simply call them `a`, `b`, and `c`, will take on four values.
+
+```python
+# the parameterized form of `test_inequality_unparameterized`
+@pytest.mark.parametrize("a, b, c", [(1, 2, 3), (4, 5, 6), (7, 8, 9), (10, 11, 12)])
+def test_inequality(a, b, c):
+    assert a < b < c
+```
+<!-- #endregion -->
+
+<div class="alert alert-warning"> 
+
+**Note**
+
+The formatting for multi-parameter tests can quickly become unwieldy.
+It isn't always obvious where one should introduce line breaks and indentations to improve readability.
+This is a place where the ["black" auto-formatter](https://black.readthedocs.io/en/stable/) really shines!
+Black will make all of these formatting decisions for us - we can write our parameterized tests as haphazardly as we like and simply run black to format our code.
+</div>
+
+
+
+<div class="alert alert-info"> 
+
+**Reading Comprehension: Parameterizing Tests**
+
+Rewrite `test_count_vowels_basic` as a parameterized test with the parameters: `input_string`, `include_y`, and `expected_count`.
+
+Rewrite `test_merge_max_mappings` as a parameterized test with the parameters: `dict_a`, `dict_b`, and `expected_merged`.
+
+Before rerunning `test_basic_functions.py` predict how many distinct test cases will be reported by pytest. 
+
+</div>
+
+
+<!-- #region -->
+Finally, you can apply multiple parameterizing decorators to a test so that pytest will run _all combinations of the respective parameter values_.
+
+```python
+# testing all combinations of `x` and `y`
+@pytest.mark.parametrize("x", [0, 1, 2])
+@pytest.mark.parametrize("y", [10, 20])
+def test_all_combinations(x, y):
+    # will run:
+    # x=0 y=10
+    # x=0 y=20
+    # x=1 y=10
+    # x=1 y=20
+    # x=2 y=10
+    # x=2 y=20
+    pass
+```
+<!-- #endregion -->
+
+### Fixtures
```

Comment:
I'm... not a big fan of fixtures.  Obviously they're very useful, and if you're using pytest you'll need to touch them eventually, but imo:

- if you're using fixtures to manage state (like a clean tmpdir), you should probably write a context manager instead.  Works for non-test code too!
- many parametrized fixtures should be replaced with Hypothesis strategies.  They were a huge advance over previous tooling, but we have another now :smile: 

They're still basically indispensable at large enough scale, when you can e.g. have your tests automatically scheduled for efficient use of database fixtures, but for this kind of tutorial I'd leave them out entirely.

Reply:
So I have been thinking about this more, and I have decided to keep this subsection in because:
 - I think that it provides significantly clearer, and more-succinct insight into what a test fixture is than do the pytest docs.
 - It provides the reader with the valuable wisdom: *your tests should not leave residue*, along with: *here is the one of the few useful fixtures that you probably want to use, and here is how you use it*